confluent_kafka/schema_registry/schema_registry_client.py

```diff
@@ -18,133 +18,258 @@
+
+    def add_schema(self, schema_id, schema):
+        """
+        Add a Schema identified by schema_id to the cache.
+
+        Args:
+            schema_id (int): Schema's registration id
+            schema (Schema): Schema instance
+
+        Returns:
+            int: The schema_id
+        """
+        # Don't overwrite existing keys
+        schema = self.get_schema(schema_id)
+        if schema is not None:
+            return
+
+        with self.lock:
+            # Pessimistic
+            if schema_id not in self.schema_id_index:
+                self.schema_id_index[schema_id] = schema
+
+    def get_schema(self, schema_id):
+        """
+        Get the schema instance associated  with schema_id from the cache.
+
+        Args:
+            schema_id (int): Id used to identify a schema
+        Returns:
+            Schema: The schema if known; else None
+
+        """
+        return self.schema_id_index.get(schema_id, None)
+
+    def add_schema_id(self, schema, schema_id):
+        """
+        Add this Schema's id to the cache.
+
+        Args:
+            schema (Schema): Schema associated with schema_id
+            schema_id (int): schema_id to cache
+        """
+
+        schema_id = self.get_schema_id(schema)
+        if schema_id is not None:
+            return
+
+        with self.lock:
+            # Pessimistic
+            if schema not in self.schema_index:
+                self.schema_index[schema] = schema_id
+
+    def get_schema_id(self, schema):
         """
-        POST /subjects/(string: subject)/versions
+        Get the schema_id associated with this schema
 
+        Args:
+            schema (Schema): The schema associated with this schema_id
+
+        Returns:
+            int: Schema ID if known; else None
+
+        """
+        return self.schema_index.get(schema, None)
+
+
+class SchemaRegistryClient(object):
+    """
+    A client that talks to a Schema Registry over HTTP
+
+    See Also:
+        http://confluent.io/docs/current/schema-registry/docs/intro.html
+
+    Arguments:
+        conf (SchemaRegistryConfig): Schema Registry config object
+    """
+
+    def __init__(self, conf):
+        self._rest_client = _RestClient(conf)
+        self.cache = _SchemaCache()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        if self._rest_client is not None:
+            self._rest_client.close()
+
+    def register_schema(self, subject_name, schema):
+        """
         Registers a schema under ``subject_name``.
         Results are cached in the ``schema`` so subsequent requests will not
         require another trip the Schema Registry.
 
         Args:
-            - subject_name (str): subject to register a schema under
-            - schema (Schema): Schema (in JSON string format) to register.
+            subject_name (str): subject to register a schema under
+            schema (Schema): Schema (in JSON string format) to register.
```

Comment:
This is confusing: is it supposed to be a JSON string schema definition, or a Schema object instance?

Reply:
it should be schema object instance